### PR TITLE
Cmake + workflows

### PR DESCRIPTION
**Pull request by wkpark**
CmakeList.txt + github workflows action

* CMake by @Jamezo97
* github workflows actions - support ubuntu and windows (windows disabled for now)
* cu11.8, cu12.1 + py310, py311 build matrix

Please see ~~https://github.com/wkpark/bitsandbytes/actions/runs/7166014267~~
~~https://github.com/wkpark/bitsandbytes/actions/runs/7167362163~~
https://github.com/wkpark/bitsandbytes/actions/runs/7648577944

### CODE REVIEW

**Comment by TimDettmers** [comment]
Thank you! We are planning to switch to make, and this is a great PR!

I do not have an overview of how this is different or could benefit from #220 and #229. Could you please have a look and discuss the differences and if we could benefit from some of the old code. Thank you!

File: .github/workflows/cmake.yml
Comment (rickardp) [inline .github/workflows/cmake.yml:19]: I would recommend building only the native library here using a build matrix, then a separate job to build the Python binary wheels.

File: CMakeLists.txt
Comment (rickardp) [inline CMakeLists.txt:123]: As an alternative, you can do something like

```
if(MSVC)
set_target_properties(bitsandbytes PROPERTIES LIBRARY_OUTPUT_DIRECTORY_RELEASE bitsandbytes)
set_target_properties(bitsandbytes PROPERTIES LIBRARY_OUTPUT_DIRECTORY_DEBUG bitsandbytes)
set_target_properties(bitsandbytes PROPERTIES RUNTIME_OUTPUT_DIRECTORY_RELEASE bitsandbytes)
set_target_properties(bitsandbytes PROPERTIES RUNTIME_OUTPUT_DIRECTORY_DEBUG bitsandbytes)
endif()

```

Possibly more cmake-canonical, not sure which is the best

**Comment by rickardp** [comment]
Is this PR complete? I remember there being some Windows-specific stuff needed, also some things for Linux/Arm64 that required specific handling.

**Comment by wkpark** [comment]
> Is this PR complete? I remember there being some Windows-specific stuff needed, also some things for Linux/Arm64 that required specific handling. Unfortunately the original PRs (#220 etc) cannot be viewed because of force pushes. I'll see if I can dig up a relevant diff

Windows specific stuff extracted on PR #873 and #876
See also https://github.com/wkpark/bitsandbytes/actions (several times compiled, partial squashed rebased blah blah..)

(and pytest stuff is not fixed/tested much since I have no sm_75 capable GPU.)

**Comment by rickardp** [comment]
> > Is this PR complete? I remember there being some Windows-specific stuff needed, also some things for Linux/Arm64 that required specific handling. Unfortunately the original PRs (#220 etc) cannot be viewed because of force pushes. I'll see if I can dig up a relevant diff
> 
> Windows specific stuff extracted on PR #873 and #876 See also https://github.com/wkpark/bitsandbytes/actions (several times compiled, partial squashed rebased blah blah..)
> 
> (and pytest stuff is not fixed/tested much since I have no sm_75 capable GPU.)

I don't have a Windows machine at hand right now, but there are also specific compiler flags that MSVC needs. I think this is what PR #220 did.

I spend a few minutes to rebase and create a draft PR in #947 just for reference, which should compile on all platforms. Maybe it's worth just comparing notes on those?

**Comment by niclimcy** [comment]
I am not sure if current versions of bitsandbytes depend on pthread but attempt #220 imports pthread-win32 is a library as pthread isn’t supported on windows either.

**Comment by wkpark** [comment]
rebased to test the current action workflows.
https://github.com/wkpark/bitsandbytes/actions/runs/7700400727/job/20984178019 (with minimal windows fix #876)

**Review by younesbelkada — APPROVED** [review]
Thanks a lot @wkpark 
I am very much in favor of this PR as it currently
1- tests bnb compilation on windowns AND ubuntu on each commit on main + on each PR to make sure future PRs will not break bnb compilation
2- makes bnb compilation easier through Cmake
3- Once we merge this PR https://github.com/TimDettmers/bitsandbytes/pull/876 should be closed
I will let @Titus-von-Koeller give a final decision on this - I assume the classic `make xxx` commands will still work - @wkpark once this gets merged, would you be happy to update the way we build bnb on our docker images to leverage CMake ? We do it exactly here: https://github.com/huggingface/peft/blob/189a9a666dc90f335b6a4db77f4b856b6a8bbdfd/docker/peft-gpu-bnb-source/Dockerfile#L60 and here: https://github.com/huggingface/peft/blob/189a9a666dc90f335b6a4db77f4b856b6a8bbdfd/docker/peft-gpu-bnb-latest/Dockerfile#L60

File: .github/workflows/cmake.yml
Comment (akx) [inline .github/workflows/cmake.yml:29]:
```suggestion
    - uses: actions/checkout@v4
```

File: .github/workflows/cmake.yml
Comment (akx) [inline .github/workflows/cmake.yml:31]:
```suggestion
      uses: actions/setup-python@v5
```

File: .github/workflows/cmake.yml
Comment (akx) [inline .github/workflows/cmake.yml:67]: Should this step then be guarded with a suitable `if:`?

File: .github/workflows/cmake.yml
Comment (akx) [inline .github/workflows/cmake.yml:152]: Calling `setup.py` directly is discouraged (see discussion in #949).

File: setup.py
Comment (akx) [inline setup.py:39]: Is there actually a Python extension module by the name of `bitsandbytes`?  If not, I don't think this is correct?

Comment (wkpark) [reply setup.py:39]: > Is there actually a Python extension module by the name of `bitsandbytes`? If not, I don't think this is correct?

there is no actual extension module named `bitsandbytes`, but how do we distinguish a linux bin and a win bin?
with this fix, we can obtain both `bitsandbytes-0.42.0-cp310-cp310-win_amd64.whl` and `bitsandbytes-0.42.0-cp310-cp310-linux_x86_64..whl`. without this fix, we can get only one `bitsandbytes-*-py3-none-any.whl` wheel.

File: .github/workflows/cmake.yml
Comment (wkpark) [reply .github/workflows/cmake.yml:67]: I mean cuda-toolkits does not work as expected. (outdated comment)

**Comment by wkpark** [comment]
> @TimDettmers says:
> Thank you! We are planning to switch to make, and this is a great PR!
> 
> I do not have an overview of how this is different or could benefit from #220 and #229. Could you please have a look and discuss the differences and if we could benefit from some of the old code. Thank you!

PR #220 by @niclimcy and PR #229 by @acpopescu are almost identical with PR #788 @Jamezo97
(some PRs are already closed.)
so I've cleaned up some unrelated stuff, cherry picked/partial squashed and extracted into minimal subsets of several PRs.

- pthread vs std::thread -> make pthread optional. : win32 fix PR
- do not rename `pythonInterface.cpp` : win32 fix PR
- AVX related : win32 fix PR
- cmake -> extract PR
- lib detection -> extract PR
- documents and trivial fixes -> drop

See also:
- https://github.com/TimDettmers/bitsandbytes/pull/220/commits/4c1d3089cc8d1f1644f334b18402a45c1ff65a44
- https://github.com/TimDettmers/bitsandbytes/pull/229/commits/68da6d806f33497013e1211d1040840768ff75ea

File: setup.py
Comment (akx) [reply setup.py:39]: Right, so the intent is to signal to the build machinery that there is native code in here and the wheel should be tagged as such 😄 

That wasn't clear from the (lack of 😉) comment or the commit message "add ext_modules" :)

Let's at least make the intent clear here so whoever will port this to a non-setup.py build knows what's up 😄 

```suggestion
    # HACK: pretend we have a native extension module so the wheel is tagged 
    #       correctly with a platform tag (e.g. `-linux_x86_64.whl`).
    ext_modules=[Extension("bitsandbytes", sources=[], language="c")],
```

**Comment by wkpark** [comment]
Partially squashed, updated.

File: .github/workflows/cmake.yml
Comment (matthewdouglas) [inline .github/workflows/cmake.yml:111]: Is it intentional to drop Maxwell here?

Comment (wkpark) [reply .github/workflows/cmake.yml:111]: > Is it intentional to drop Maxwell here?

that's not intended. I just removed some computes to reduce compile time a bit.

File: .github/workflows/cmake.yml
Comment (younesbelkada) [inline .github/workflows/cmake.yml:18]:
```suggestion
        os: [ubuntu-latest, windows-latest]
```

**Review by younesbelkada — APPROVED** [review]
Thanks @wkpark for your great work ! as #876 being merged could you re-enable windows build in the workflow + rebase with main?

**Comment by younesbelkada** [comment]
@wkpark could you also update the documentation and the main reamde with instructions on how to build bnb with windows?

**Comment by younesbelkada** [comment]
For the docs you can start filling some content here: https://github.com/TimDettmers/bitsandbytes/blob/main/docs/source/installation.mdx

File: CMakeLists.txt
Comment (wkpark) [reply CMakeLists.txt:127]: thank you for your information. I've missed `LIBRARY_OUTPUT_DIRECTORY` property!

**Comment by wkpark** [comment]
- [x] fixed mistake d97700af6 in PR #876
- [x] add windows build matrix

**Comment by younesbelkada** [comment]
@wkpark Thanks a lot ! let me know if you need any help for https://github.com/TimDettmers/bitsandbytes/pull/908#issuecomment-1920428847 - could you just share how to build bnb from source using a windows machine, i.e, the exact commands and I'll push on your branch

**Comment by wkpark** [comment]
> For the docs you can start filling some content here: https://github.com/TimDettmers/bitsandbytes/blob/main/docs/source/installation.mdx

this is just a note for windows user who wanted to build from source.

1. git clone
2. `cmake -B` ***build*** `-DBUILD_CUDA=ON -S .`
3. `cmake --build` ***build*** `--config Release`
4. `python -m build --wheel` (build python wheel using `build` module.) then you can get `dist/*.whl`

**Comment by younesbelkada** [comment]
@wkpark check out https://moon-ci-docs.huggingface.co/docs/bitsandbytes/pr_908/en/installation now there is a nice table that showcases how to install bnb: 

Screenshot 2024-02-01 at 16 16 56

Screenshot 2024-02-01 at 16 17 17

File: docs/source/installation.mdx
Comment (younesbelkada) [inline docs/source/installation.mdx:40]:
```suggestion
Big thanks to [wkpark](https://github.com/wkpark), [Jamezo97](https://github.com/Jamezo97), [rickardp](https://github.com/rickardp), [akx](https://github.com/akx) for their amazing contributions to make bitsandbytes compatible with Windows.
```

**Comment by younesbelkada** [comment]
Thanks again for all your awesome contributions ! @rickardp @wkpark @akx @Jamezo97 !

File: CMakeLists.txt
Comment (Neumann-A) [inline CMakeLists.txt:112]: Should have used targets from https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html . This now assumes that the libraries are on the library search path

Comment (akx) [reply CMakeLists.txt:112]: All of this will be tweaked soon anyway, don't worry – see #949 for the follow-up.

**Pull request merged**
